searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.h

```diff
@@ -455,7 +401,70 @@ getIterators(bool shouldApplyRangeLimit)
+{
+    // The following initial constants are derived after running parts of
+    // the range search performance test with 10M documents on an Apple M1 Pro with 32 GB memory.
+    // This code was compiled with two different behaviors:
+    //   1) 'filter matching' (never use posting lists).
+    //   2) 'posting list matching' (always use posting lists).
+    // https://github.com/vespa-engine/system-test/tree/master/tests/performance/range_search
+    //
+    // The following test case was used to establish the baseline cost of producing different number of hits as cheap as possible:
+    // range_hits_ratio=[1, 10, 50, 100, 200, 500], values_in_range=1, fast_search=true, filter_hits_ratio=0.
+    // The 6 range queries end up using a single posting list that produces the following number of hits: [10k, 100k, 500k, 1M, 2M, 5M]
+    // Avg query latency (ms) results: [5.43, 8.56, 11.68, 14.68, 22.77, 42.88]
+    //
+    // Then the following test case was executed for both 1) 'filter matching' and 2) 'posting list matching':
+    // range_hits_ratio=[1, 10, 50, 100, 200, 500], values_in_range=100, fast_search=true, filter_hits_ratio=0.
+    // Avg query latency (ms) results:
+    //   1) 'filter matching':       [47.52, 51.06, 59.68, 79.3, 118.7, 145.26]
+    //   2) 'posting list matching': [4.79, 11.6, 13.54, 20.24, 32.65, 67.28]
+    //
+    // For 1) 'filter matching' we use the result from range_hits_ratio=1 (10k hits) compared to the baseline
+    // to calculate the cost per document (in ns) to do filter matching: 1M*(47.52-5.43)/10M = 4.2
+    //
+    // For 2) 'posting list matching' we use the results from range_hits_ratio=[50, 100, 200, 500] compared to the baseline
+    // to calculate the average cost per hit (in ns) when merging the 100 posting lists:
+    //   1M*(13.54-11.68)/500k = 3.7
+    //   1M*(20.24-14.68)/1M = 5.6
+    //   1M*(32.65-22.77)/2M = 4.9
+    //   1M*(67.28-42.88)/5M = 4.9
+    //
+    // The average is 4.8.
+
+    constexpr float filtering_match_constant = 4.2;
+    constexpr float posting_list_merge_constant = 4.8;
+
+    uint32_t exp_doc_hits = this->_docIdLimit * info.hitRate();
+    float avg_values_per_document = static_cast<float>(this->_numValues) / static_cast<float>(this->_docIdLimit);
+    float filtering_cost = exp_doc_hits * avg_values_per_document * filtering_match_constant;
+    float posting_list_cost = this->estimated_hits_in_range() * posting_list_merge_constant;
+    return posting_list_cost < filtering_cost;
+}
 
+template <typename BaseSC, typename AttrT, typename DataT>
+size_t
+NumericPostingSearchContext<BaseSC, AttrT, DataT>::calc_estimated_hits_in_range() const
+{
+    size_t exact_sum = 0;
+    size_t estimated_sum = 0;
+    uint32_t count = 0;
+    constexpr uint32_t max_posting_lists_to_count = 1000;
+    for (auto it = this->_lowerDictItr; it != this->_upperDictItr; ++it) {
+        if (count >= max_posting_lists_to_count) {
```

Comment:
I find it easier to read and understand if you add (count < max_posting_lists_to_count) as criteria in the for loop,
and have if branch after the for loop. Then you don't need the break. either. It makes it more clear what happens in the for loop.

Reply:
Fixed